### DISCUSSION TITLE
--key not 16/24/32 length is secured and what happens ?

### DISCUSSION THREAD

**Discussion by gaamaaresosa** [pyinstaller/pyinstaller#7576] [Features & Development]
I know on v6x the --key is going to be removed.
Before that please let me know what happens when I set only 15/25/33 --key length on pyinstaller build ?
Do you ignore that wrong length --key or accept and build now ?
Or cryto is not done because of wrong --key length ?

Recently I am into analyzing security issues with pyinstaller to make more secured.
I tried setting wrong length --key size (15/25/33) and built.
It was success building...

Then I used "pyinstxtractor" and other tools to extract the pyinstaller protected binary.
I think I was failed to extract it because my --key length was incorrect for Crypto.Cipher.
But 'strings ' command could extract all the strings used on the code.

Whats happening exactly here ?
It is secure to use wrong length --key size (15/25/33)  on pyinstaller ?

**Comment by rokm** [comment]
The given key is zero-padded or truncated to the length of 16 - see the build-time part of the code

https://github.com/pyinstaller/pyinstaller/blob/5c9f3e65affd43faa087da3ef6ddea6fa5947f98/PyInstaller/archive/pyz_crypto.py#L24-L39

or the run-time part:

https://github.com/pyinstaller/pyinstaller/blob/5c9f3e65affd43faa087da3ef6ddea6fa5947f98/PyInstaller/loader/pyimod01_archive.py#L47-L64

> Then I used "pyinstxtractor" and other tools to extract the pyinstaller protected binary.

I don't think `pyinstxtractor` supports byte-encrypted PYZ, so your "test" is likely meaningless.

>  But 'strings ' command could extract all the strings used on the code.

Strings from where? From byte-compiled python modules or from data files / binaries? Because bytecode-encryption is applied only to collected .pyc modules\*, and nothing else...

[\*] And even that covers only .pyc modules that are collected into PYZ archive, which is most of them (except modules needed at the bootstrap (which are in `base_library.zip`), pyinstaller's bootstrap scripts and modules, and your entry-point script).

>  It is secure to use wrong length --key size (15/25/33) on pyinstaller ?

It makes no difference from security viewpoint. That is, the built-in bytecode encryption is trivial to overcome no matter key size you use.

**Reply by gaamaaresosa** [reply]
Thanks for your explanation rokm !
I got it...
